gtest/src/mailbox.rs

```diff
@@ -0,0 +1,262 @@
+            program_id,
+            manager,
+        }
+    }
+
+    pub fn contains<T: Into<Log> + Clone>(&self, log: &T) -> bool {
+        let log: Log = log.clone().into();
+        match self.manager.borrow().mailbox.get(&self.program_id) {
+            None => {
+                self.manager
+                    .borrow_mut()
+                    .mailbox
+                    .insert(self.program_id, Vec::default());
+                false
+            }
+            Some(mailbox) => mailbox.iter().any(|message| log.eq(message)),
+        }
+    }
+
+    pub fn take_message<T: Into<Log> + Clone>(&self, log: T) -> MessageReplier {
+        let log: Log = log.into();
+        let index = match self.manager.borrow().mailbox.get(&self.program_id) {
+            None => {
+                self.manager
+                    .borrow_mut()
+                    .mailbox
+                    .insert(self.program_id, Vec::default());
+                panic!("No message that satisfies log");
+            }
+            Some(mailbox) => mailbox
+                .iter()
+                .position(|message| log.eq(message))
+                .expect("No message that satisfies log"),
+        };
+
+        let taken_message = self
+            .manager
+            .borrow_mut()
+            .mailbox
+            .get_mut(&self.program_id)
+            .unwrap()
+            .remove(index);
+
+        MessageReplier::new(taken_message, self.manager)
+    }
+
+    pub fn reply(&self, log: Log, payload: impl Encode, value: u128) -> RunResult {
+        self.take_message(log).reply(payload, value)
+    }
+
+    pub fn reply_bytes(&self, log: Log, raw_payload: impl AsRef<[u8]>, value: u128) -> RunResult {
+        self.take_message(log).reply_bytes(raw_payload, value)
+    }
+}
+
+pub struct MessageReplier<'a> {
+    log: CoreLog,
+    manager: &'a RefCell<ExtManager>,
+}
+
+impl<'a> MessageReplier<'a> {
+    pub(crate) fn new(
+        message: StoredMessage,
+        manager: &'a RefCell<ExtManager>,
+    ) -> MessageReplier<'a> {
+        MessageReplier {
+            log: message.into(),
+            manager,
+        }
+    }
+
+    pub fn reply(&self, payload: impl Encode, value: u128) -> RunResult {
+        self.reply_bytes(payload.encode(), value)
+    }
+
+    pub fn reply_bytes(&self, raw_payload: impl AsRef<[u8]>, value: u128) -> RunResult {
+        let message = Message::new(
+            MessageId::from(self.manager.borrow_mut().fetch_inc_message_nonce()),
+            self.log.destination(),
+            self.log.source(),
+            raw_payload.as_ref().to_vec(),
+            None,
+            value,
+            self.log
+                .exit_code()
+                .map(|exit_code| (self.log.id(), exit_code)),
+        );
+
+        self.manager
+            .borrow_mut()
+            .run_dispatch(Dispatch::new(DispatchKind::Reply, message))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::program::ProgramIdWrapper;
```

Comment:
Also it would be useful to show panic cases. Rust test-mechanism has `#[should_panic(expected="panic message"]` proc macro for it